lib/stages/mergeBeamBuffer.cpp

```diff
@@ -0,0 +1,130 @@
+    // when the ID reaches the num_frame, we need something something to keep track
+    // on the incoming frame number
+    uint32_t in_frame_counter = 0;
+    
+    while (!stop_thread) {
+        // Get an input buffer, This call is blocking!
+        in_frame = wait_for_full_frame(in_buf, unique_name.c_str(), in_buffer_ID);
+        if (in_frame == nullptr)
+            break;
+
+	BeamMetadata* in_metadata = (BeamMetadata*)get_metadata(in_buf, in_buffer_ID);
+        
+        // Compute the frame index in the outbuffer
+        uint32_t sub_frame_index = in_frame_counter % _sub_frames_per_merged_frame;
+        // compute the size of the sub_frame of the out_frame
+        uint32_t FreqIDBeamMeta_size = sizeof(FreqIDBeamMetadata);
+        uint32_t sub_frame_size = FreqIDBeamMeta_size + in_buf -> frame_size;
+        
+	/// Get a new merged frame for a new set of sub frames. 
+	if (sub_frame_index == 0){
+            out_frame = (uint8_t*)wait_for_empty_frame(out_buf, unique_name.c_str(), out_buffer_ID);
+            if (out_frame == nullptr)
+                break;
+	    allocate_new_metadata_object(out_buf, out_buffer_ID);
+            MergedBeamMetadata* out_metadata = (MergedBeamMetadata*)get_metadata(out_buf, out_buffer_ID);
+	    // populate the merged_buffer/out_buffer frame header.
+	    out_metadata -> sub_frame_pre_frame = _sub_frames_per_merged_frame;
+            out_metadata -> sub_frame_metadata_size = FreqIDBeamMeta_size;
+            out_metadata -> sub_frame_data_size = in_buf -> frame_size;
+	}
+
+	// Get the start position of a sub_out_frame's meatadata
+	// Make sure out_frame is pointed to the right place.
+        if (out_frame == nullptr)
+            break;
+	uint32_t out_frame_metadata_pos = sub_frame_size * sub_frame_index;
+	// Get the memory block for the metadata of sub_out_frame.
+	FreqIDBeamMetadata* sub_frame_metadata = (FreqIDBeamMetadata *)&out_frame[out_frame_metadata_pos];
+	// Copy the in_frame metadata the the out frame's subframe metadata
+	// Using memory copy here will give a warning.
+	sub_frame_metadata -> fpga_seq_start = in_metadata -> fpga_seq_start;
+	sub_frame_metadata -> ctime = in_metadata -> ctime;
+	sub_frame_metadata -> stream_id = in_metadata -> stream_id;
+	sub_frame_metadata -> dataset_id = in_metadata -> dataset_id;
+	sub_frame_metadata -> beam_number = in_metadata -> beam_number;
+	sub_frame_metadata -> ra = in_metadata -> ra;
+	sub_frame_metadata -> dec = in_metadata -> dec;
+	sub_frame_metadata -> scaling = in_metadata -> scaling;
+        
+	// Add the frequency id into the metadata, for chime one stream has on frequency.
+	// TODO Have the output buffer handle the stream with mulitple frequency
+	const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+
+        for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+	    // TODO this does not handling the stream with mulitple frequencies. since at right now, 
```

Comment:
This is generally best handled by encoding the variable `num_freq_per_stream` into your metadata, and then having a variable length section after the metadata, but before the voltage data, for storing each bin.  This does make the indexing more complex, but its really the only way to account for a variable length in the metadata.